pkg/session/session.go

```diff
@@ -3,68 +3,73 @@ package session
+// on that information, Session extends it and provides harder rules that
+// arise from configuration and other static heuristics.
 type Session struct {
 	State
-	config Config
+	config *Config
 }
 
-func NewSession(config Config, withState bool) *Session {
+// NewSession constructs a new Session with the given config and with or
+// without ongoing dynamic data collection. When withState is false, no
+// dynamic data will be collected and decisions will be made solely based
+// on the config.
+func NewSession(config *Config, withState bool) *Session {
+	if config == nil {
+		config = DefaultConfig()
+	}
 	var state State
 	if withState {
-		state = newSpTracker(nil)
+		state = NewSessionState(config)
 	} else {
-		state = &nilstate{}
+		state = nilstate{}
 	}
-	return &Session{
-		State:  state,
-		config: config,
+	if config == nil {
+		config = &Config{}
 	}
+	return &Session{state, config}
 }
-func (s *Session) GetStorageProviderTimeout(storageProviderId peer.ID) time.Duration {
-	return s.config.getMinerConfig(storageProviderId).RetrievalTimeout
+
+// GetStorageProviderConcurrency returns the per-retrieval timeout from the
+// RetrievalTimeout configuration option.
+func (session *Session) GetStorageProviderTimeout(storageProviderId peer.ID) time.Duration {
+	return session.config.getProviderConfig(storageProviderId).RetrievalTimeout
 }
 
-// isAcceptableStorageProvider checks whether the storage provider in question
-// is acceptable as a retrieval candidate. It checks the blacklists and
-// whitelists, the miner monitor for failures and whether we are already at
-// concurrency limit for this SP.
-func (s *Session) FilterIndexerCandidate(candidate types.RetrievalCandidate) (bool, types.RetrievalCandidate) {
+// FilterIndexerCandidate filters out protocols that are not acceptable for
+// the given candidate. It returns a bool indicating whether the candidate
+// should be considered at all, and a new candidate with the filtered
+// protocols.
+func (session *Session) FilterIndexerCandidate(candidate types.RetrievalCandidate) (bool, types.RetrievalCandidate) {
+	if !session.isAcceptableCandidate(candidate.MinerPeer.ID) {
+		return false, types.RetrievalCandidate{
+			MinerPeer: candidate.MinerPeer,
+			RootCid:   candidate.RootCid,
+			Metadata:  metadata.Default.New(),
+		}
+	}
+
 	protocols := candidate.Metadata.Protocols()
 	newProtocolMetadata := make([]metadata.Protocol, 0, len(protocols))
+	seen := make(map[multicodec.Code]struct{})
 	for _, protocol := range protocols {
-		includeProtocol := true
-		switch protocol {
-		case multicodec.TransportGraphsyncFilecoinv1:
-			includeProtocol = s.isAcceptableGraphsyncCandidate(candidate.MinerPeer.ID)
-		case multicodec.TransportBitswap:
-			includeProtocol = s.isAcceptableBitswapCandidate(candidate.MinerPeer.ID)
-		case multicodec.TransportIpfsGatewayHttp:
-			includeProtocol = s.isAcceptableHttpCandidate(candidate.MinerPeer.ID)
+		if _, ok := seen[protocol]; ok {
```

Comment:
probably a reasonable add. i hope people aren't making adds like this. Is there a scenario where we get a graphsync twice with different parameters?

Reply:
I think that's a second result.

Reply:
https://github.com/ipni/go-libipni/blob/1bd7b3f90d2a849c3449742947e86d78e66508ab/metadata/metadata.go#L118-L119

😬

I don't believe this is validated either on the consuming (indexer) side or the producer (provider), since it's possible (i.e. it's a `[]Protocol` on the struct), then I guessed it'd be safer than not bothering and was easy enough.

Reply:
FYI @masih just so it's on your radar that someone's thinking about the potential problems here. Not a big deal because we can, and probably should, check, but we were already caught not filtering out duplicate entries with different context IDs in an earlier form of Lassie and this is another potential footgun for consumers.

i.e. a consumer of indexer data needs to be clear about the scope of the information they care about and take steps to de-duplicate to the appropriate degree to solve that. We're checking double-protocols for individual entries, elsewhere (also in this PR btw cause it moved) we're deduplicating candidates and merging metadata to make sure we always have the "best" metadata for a provider (i.e. if they say they have FastRetrieval in one but not in another, we'll just record that they do regardless of order that we see that).